rclrs/src/timer.rs

```diff
@@ -0,0 +1,812 @@
+            last_elapse: Mutex::new(Duration::ZERO),
+            lifecycle: Mutex::default(),
+            _ignore: Default::default(),
+        });
+
+        let (waitable, lifecycle) = Waitable::new(
+            Box::new(TimerExecutable::<Scope> {
+                timer: Arc::downgrade(&timer),
+                handle: Arc::clone(&timer.handle),
+            }),
+            Some(Arc::clone(commands.get_guard_condition())),
+        );
+
+        *timer.lifecycle.lock().unwrap() = Some(lifecycle);
+
+        commands.add_to_wait_set(waitable);
+
+        Ok(timer)
+    }
+
+    /// Force the timer to be called, even if it is not ready to be triggered yet.
+    /// We could consider making this public, but the behavior may confuse users.
+    fn call(self: &Arc<Self>, any_payload: &mut dyn Any) -> Result<(), RclrsError> {
+        // Keep track of the time elapsed since the last call. We need to run
+        // this before we trigger rcl_call.
+        let last_elapse = self.time_since_last_call().unwrap_or(Duration::ZERO);
+        *self.last_elapse.lock().unwrap() = last_elapse;
+
+        if let Err(err) = self.rcl_call() {
+            log_error!("timer", "Unable to call timer: {err:?}",);
+        }
+
+        let Some(callback) = self.callback.lock().unwrap().take() else {
+            log_error!(
+                "timer".once(),
+                "Timer is missing its callback information. This should not \
+                be possible, please report it to the maintainers of rclrs.",
+            );
+            return Ok(());
+        };
+
+        let Some(payload) = any_payload.downcast_mut::<Scope::Payload>() else {
+            return Err(RclrsError::InvalidPayload {
+                expected: std::any::TypeId::of::<Scope::Payload>(),
+                received: (*any_payload).type_id(),
+            });
+        };
+
+        match callback {
+            AnyTimerCallback::Repeating(mut callback) => {
+                callback(payload, self);
+                self.restore_callback(AnyTimerCallback::Repeating(callback).into());
+            }
+            AnyTimerCallback::OneShot(callback) => {
+                callback(payload, self);
+                self.restore_callback(AnyTimerCallback::Inert);
+            }
+            AnyTimerCallback::Inert => {
+                self.restore_callback(AnyTimerCallback::Inert);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Updates the state of the rcl_timer to know that it has been called. This
+    /// should only be called by [`Self::call`].
+    ///
+    /// The callback held by the rcl_timer is null because we store the callback
+    /// in the [`Timer`] struct. This means there are no side-effects to this
+    /// except to keep track of when the timer has been called.
+    fn rcl_call(&self) -> Result<(), RclrsError> {
+        let mut rcl_timer = self.handle.rcl_timer.lock().unwrap();
+        unsafe { rcl_timer_call(&mut *rcl_timer) }.ok()
```

Comment:
Shouldn't this have a safety comment?

Reply:
[a589997](https://github.com/ros2-rust/ros2_rust/pull/480/commits/a589997b24b07b7a934b7431604f25d8e82a815a)